### PR TITLE
Read KAZR ingest

### PR DESCRIPTION
**Pull request by kirknorth**
Adding the ability to read KAZR ingest (e.g., a1 data). These file types are characterized as NetCDF files that are not fully CF/Radial in terms of their convention. Nonetheless it borrows heavily from the existing CF/Radial module. I haven't looked at other data streams (e.g., 915-MHz RWP), but if they have an ingest similar to the KAZR, then read_kazr can likely be renamed to something more generic...

In terms of testing, I have only confirmed that I am indeed able to create a radar object from KAZR data. I have not yet tried to use it in any of the other modules (e.g., `correct` or `graph`).

### CODE REVIEW

**Comment by kirknorth** [comment]
This PR addresses Issue #270.

**Comment by kirknorth** [comment]
I now see that the `instrument_parameters` dictionary is by default set to `None`, however this information is readily available in the initial NetCDF (e.g., Nyquist velocity) and thus we should populate this accordingly.

**Comment by kirknorth** [comment]
Seems to be working fine with the `graph` module. FYI I'm not plotting the whole time range, just the first 5000 rays.

sgpkazrgec1 a1 20110511 000000

**Comment by scollis** [comment]
Very pretty.. Thanks for the PR!
DO you know if WACR is in the same format?

**Comment by kirknorth** [comment]
Not sure @scollis, but like I said in one of the comments, if these other instruments have the same format, then this module is indeed more generic and can be renamed.

File: pyart/aux_io/arm_vpt.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        shape = (len(ncvars['time']), len(ncvars['range']))
+        ray_n_gates = ncvars['ray_n_gates'][:]
+        ray_start_index = ncvars['ray_start_index'][:]
+        for dic in fields.values():
+            cfradial._unpack_variable_gate_field_dic(
+                dic, shape, ray_n_gates, ray_start_index)
+
+    # 4.5 instrument_parameters sub-convention -> instrument_parameters dict
+
+    # 4.6 radar_parameters sub-convention -> instrument_parameters dict
```
Comment (kirknorth) [inline pyart/aux_io/arm_vpt.py]: This section needs some work!

**Comment by kirknorth** [comment]
@jjhelmus I'm testing the new velocity correction routines on KAZR data. Is there anything inherent in the phase algorithm that shouldn't allow me to set `unwrap_unit = 'sweep'` for this VPT data? A sweep for VPT data is just the whole time-height record. It allows me to use `unwrap_unit = 'ray'` but the results are very poor. The results from the region based algorithm are quite nice. Here's my [notebook](http://nbviewer.ipython.org/github/kirknorth/lab-notebooks/blob/master/test/SGP_KAZR_Basic_Processing.ipynb) for reference.

**Comment by jjhelmus** [comment]
@kirknorth By default the `dealias_unwrap_phase` function verifies that the **unwrap_unit parameter** is valid.  In the case of the 'sweep' (and 'volume') this check consists of a verification that the sweep were collected sequentially, ie no jumps in azimuth/elevation angles for PPI/RHI scans.  This does not  have an option for vpt scans ([see line 276 in unwrap.py](https://github.com/ARM-DOE/pyart/blob/ead72e73ab440acff92c206fdaa7c28b2fb82a97/pyart/correct/unwrap.py#L276)) but probably should.  A check which verifies that the time coordinate does not run backwards is best, I'll add it momentarily.

For the time being this verification can be skipped by setting the **skip_checks** parameter to True which disable all verification.

**Comment by jjhelmus** [comment]
Also FYI the in RWP a1 files the data for the short and long pulse moments are interleaved and have different ranges for the two pulse length.  An separate reader which returns two (or more) Radar object would need to be written for these files.

File: pyart/aux_io/arm_vpt.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+    altitude = cfradial._ncvar_to_dict(ncvars['alt'])
+    if 'altitude_agl' in ncvars:
+        altitude_agl = cfradial._ncvar_to_dict(ncvars['altitude_agl'])
+    else:
+        altitude_agl = None
+
+    # 4.7 Sweep variables -> create atrribute dictionaries
+    # this is the section that needed the most work since the initial NetCDF
+    # file did not contain any sweep information
+    sweep_number = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: Use the pyart.common.get_metadata to retrieve a dictionary of metadata that will be consistent and can be customized by the user by adjusting the config file.

File: pyart/aux_io/arm_vpt.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+        'data': np.array([0], dtype=np.int32),
+        'long_name': 'Sweep number',
+        'standard_name': 'sweep_number',
+        'units': 'count',
+    }
+    sweep_mode = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: pyart.common.get_metadata

File: pyart/aux_io/arm_vpt.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+        'data': np.array(['vertical_pointing'], dtype=np.str),
+        'long_name': 'Sweep mode',
+        'standard_name': 'sweep_mode',
+        'units': 'unitless',
+        'comment': ('Options are: "sector", "coplane", "rhi" '
+                    '"vertical_pointing", "idle", "azimuth_surveillance" '
+                    '"elevation_surveillance", "sunscan", "pointing" '
+                    '"manual_ppi", "manual_rhi"')
+    }
+    fixed_angle = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: pyart.common.get_metadata

File: pyart/aux_io/arm_vpt.py [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+        'data': np.array([90.0], dtype=np.float32),
+        'long_name': 'Target angle for sweep',
+        'standard_name': 'target_fixed_angle',
+        'units': 'degrees',
+    }
+    sweep_start_ray_index = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: pyart.common.get_metadata

File: pyart/aux_io/arm_vpt.py [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+        'data': np.array([0], dtype=np.int32),
+        'long_name': 'Index of first ray in sweep, 0-based',
+        'standard_name': 'index_of_first_ray_in_sweep',
+        'units': 'count',
+    }
+    sweep_end_ray_index = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: pyart.common.get_metadata

File: pyart/aux_io/arm_vpt.py [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+
+    # first sweep mode determines scan_type
+    # this module is specific to vertically-pointing data
+    mode = 'vertical_pointing'
+    scan_type = 'vpt'
+
+    # 4.8 Sensor pointing variables -> create attribute dictionaries
+    # this section also required some changes since the initial NetCDF did not
+    # contain any sensor pointing variables
+    azimuth = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: pyart.common.get_metadata

**Comment by kirknorth** [comment]
@jjhelmus I assume the RWP short/long pulse mode is not supported by CF/Radial?

File: pyart/aux_io/arm_vpt.py [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+        'data': 0.0 * np.ones(ncvars['time'].size, dtype=np.float32),
+        'long_name': 'Azimuth angle relative to true north',
+        'standard_name': 'beam_azimuth_angle',
+        'units': 'degrees',
+        'axis': 'radial_azimuth_angle',
+        'comment': 'Azimuth of antenna relative to true north',
+    }
+    elevation = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: pyart.common.get_metadata

File: pyart/aux_io/arm_vpt.py [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+    else:
+        scan_rate = None
+
+    if 'antenna_transition' in ncvars:
+        antenna_transition = cfradial._ncvar_to_dict(
+            ncvars['antenna_transition'])
+    else:
+        antenna_transition = None
+
+    # 4.9 Moving platform geo-reference variables
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: I think this section can be removed.  I don't know how ARM is specifying movement on shipborne zenith pointing radars but I doubt it follow the Cf/Radial convention.

File: pyart/aux_io/arm_vpt.py [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+        pitch = _ncvar_to_dict(ncvars['pitch'])
+    else:
+        pitch = None
+    if 'georefs_applied' in ncvars:
+        georefs_applied = _ncvar_to_dict(ncvars['georefs_applied'])
+    else:
+        georefs_applied = None
+
+    # 4.10 Moments field data variables -> field attribute dictionary
+    if 'ray_n_gates' in ncvars:
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: This if clause can also be removed, ARM's zenith radar should maintain a consistent number of gates in each ray.  The else clause should always apply.

File: pyart/aux_io/arm_vpt.py [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+    fields = {}
+    for key in keys:
+        field_name = filemetadata.get_field_name(key)
+        if field_name is None:
+            if exclude_fields is not None and key in exclude_fields:
+                continue
+            field_name = key
+        fields[field_name] = cfradial._ncvar_to_dict(ncvars[key])
+
+    if 'ray_n_gates' in ncvars:
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: This can be removed.

File: pyart/aux_io/arm_vpt.py [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+        for dic in fields.values():
+            cfradial._unpack_variable_gate_field_dic(
+                dic, shape, ray_n_gates, ray_start_index)
+
+    # 4.5 instrument_parameters sub-convention -> instrument_parameters dict
+    # this section needed multiple changes and/or additions since the
+    # instrument parameters were primarily located in the global attributes
+    # this section is likely still incomplete
+    omega = float(ncobj.radar_operating_frequency.split()[0])
+    frequency = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: pyart.common.get_metadata

File: pyart/aux_io/arm_vpt.py [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+        ray_start_index = ncvars['ray_start_index'][:]
+        for dic in fields.values():
+            cfradial._unpack_variable_gate_field_dic(
+                dic, shape, ray_n_gates, ray_start_index)
+
+    # 4.5 instrument_parameters sub-convention -> instrument_parameters dict
+    # this section needed multiple changes and/or additions since the
+    # instrument parameters were primarily located in the global attributes
+    # this section is likely still incomplete
+    omega = float(ncobj.radar_operating_frequency.split()[0])
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: Should this be protected with a try or if 'radar_operating_frequency' in ncobj.ncattrs block?

File: pyart/aux_io/arm_vpt.py [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+    frequency = {
+        'data': np.array([omega / 1e9], dtype=np.float32),
+        'long_name': 'Radar operating frequency',
+        'standard_name': 'radiation_frequency',
+        'units': 'Hz',
+        'meta_group': 'instrument_parameters',
+    }
+    prt_mode = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: pyart.common.get_metadata

File: pyart/aux_io/arm_vpt.py [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+        'data': np.array(['fixed'], dtype=np.str),
+        'long_name': 'Pulsing mode',
+        'standard_name': 'transmit_pulse_mode',
+        'units': 'unitless',
+        'meta_group': 'instrument_parameters',
+        'comment': ('Options are "fixed", "staggered", "dual". Assumed '
+                    '"fixed" if missing'),
+    }
+    prf = float(ncobj.pulse_repetition_frequency.split()[0])
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: try/ if block?

File: pyart/aux_io/arm_vpt.py [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+    prt = {
+        'data': (1.0 / prf) * np.ones(ncvars['time'].size, dtype=np.float32),
+        'long_name': 'Pulse repetition time',
+        'standard_name': 'pulse_repetition_time',
+        'units': 'seconds',
+        'meta_group': 'instrument_parameters',
+        'comments': 'For staggered PRT, also see prt_ratio',
+    }
+    v_nq = float(ncobj.nyquist_velocity.split()[0])
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: try/ if block

File: pyart/aux_io/arm_vpt.py [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+        'data': np.array(['fixed'], dtype=np.str),
+        'long_name': 'Pulsing mode',
+        'standard_name': 'transmit_pulse_mode',
+        'units': 'unitless',
+        'meta_group': 'instrument_parameters',
+        'comment': ('Options are "fixed", "staggered", "dual". Assumed '
+                    '"fixed" if missing'),
+    }
+    prf = float(ncobj.pulse_repetition_frequency.split()[0])
+    prt = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: pyart.common.get_metadata

File: pyart/aux_io/arm_vpt.py [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+        'data': (1.0 / prf) * np.ones(ncvars['time'].size, dtype=np.float32),
+        'long_name': 'Pulse repetition time',
+        'standard_name': 'pulse_repetition_time',
+        'units': 'seconds',
+        'meta_group': 'instrument_parameters',
+        'comments': 'For staggered PRT, also see prt_ratio',
+    }
+    v_nq = float(ncobj.nyquist_velocity.split()[0])
+    nyquist_velocity = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: pyart.common.get_metadata

File: pyart/aux_io/arm_vpt.py [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+        'data': v_nq * np.ones(ncvars['time'].size, dtype=np.float32),
+        'long_name': 'Nyquist velocity',
+        'standard_name': 'unambiguous_doppler_velocity',
+        'units': 'meters_per_second',
+        'meta_group': 'instrument_parameters',
+    }
+    samples = int(ncobj.num_spectral_averages)
+    n_samples = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: pyart.common.get_metadata

File: pyart/aux_io/arm_vpt.py [hunk 20]
```diff
@@ -0,0 +1,351 @@
...
+    }
+    v_nq = float(ncobj.nyquist_velocity.split()[0])
+    nyquist_velocity = {
+        'data': v_nq * np.ones(ncvars['time'].size, dtype=np.float32),
+        'long_name': 'Nyquist velocity',
+        'standard_name': 'unambiguous_doppler_velocity',
+        'units': 'meters_per_second',
+        'meta_group': 'instrument_parameters',
+    }
+    samples = int(ncobj.num_spectral_averages)
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: try/ if block?

File: pyart/aux_io/arm_vpt.py [hunk 21]
```diff
@@ -0,0 +1,351 @@
...
+        'meta_group': 'instrument_parameters',
+    }
+
+    # 4.6 radar_parameters sub-convention -> instrument_parameters dict
+    # this section needed multiple changes and/or additions since the
+    # radar instrument parameters were primarily located in the global
+    # attributes
+    # this section is likely still incomplete
+
+    instrument_parameters = {
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: If a try/if block is used above then instrument_parameters should be built up sequentially with a check for an empty dictionary at the end.

File: pyart/aux_io/arm_vpt.py [hunk 22]
```diff
@@ -0,0 +1,351 @@
...
+    else:
+        rays_are_indexed = None
+    if 'ray_angle_res' in ncvars:
+        ray_angle_res = cfradial._ncvar_to_dict(ncvars['ray_angle_res'])
+    else:
+        ray_angle_res = None
+
+    # first sweep mode determines scan_type
+    # this module is specific to vertically-pointing data
+    mode = 'vertical_pointing'
```
Comment (jjhelmus) [inline pyart/aux_io/arm_vpt.py]: is this variable used elsewhere?

**Comment by jjhelmus** [comment]
@kirknorth No, Cf/Radial does not support a concept of mixed modes.  In addition the gate spacing in the RWP changes between modes so the range variable is not constant which Cf/Radial also has no method of expressing.  Scott and I am going to work with folks at NCAR to suggest some change to the format which would allow an option for variable range spacing in a single volume but nothing yet.

**Comment by jjhelmus** [comment]
This looks like a good addition to Py-ART.  Since the A stands for ARM it would be good to support as many ARM radars as possible.  I added some comment to the code which I think should be addressed before this is merged.  @kirknorth let me know if you want to address these or I can submit a PR against your branch to clean these up.

**Comment by kirknorth** [comment]
Jonathan, please go ahead and submit a PR against my branch to address
everything you need to fix. Like I said in the earlier comments, feel free
to move and/or rename anything as well.

Kirk

**Comment by jjhelmus** [comment]
Will do.  Might need to wait until next week,  I'm heading out of town today so am only putting in a half day.

**Comment by jjhelmus** [comment]
Kirk, sorry I did not get to this pull request earlier.  I submitted a [pull against your branch](https://github.com/kirknorth/pyart/pull/5) which address the concerns and resolved the merge conflicts in the branch.  It does not add support for other ARM VPT files which will require additional work.  Once you accept this I can merge this into master.  Thanks for the contribution.

**Comment by kirknorth** [comment]
No worries @jjhelmus, glad to contribute. Let me merge this on my end.

**Comment by jjhelmus** [comment]
Merging.  That is all for today, I'll be looking at the other outstanding PRs tomorrow.

**Pull request merged**